slatedb/src/dispatcher.rs

```diff
@@ -0,0 +1,813 @@
+        let error_state = WatchableOnceCell::new();
+        let fp_registry = Arc::new(FailPointRegistry::default());
+        let mut dispatcher = MessageDispatcher::new_with_fp_registry(
+            Box::new(handler),
+            rx,
+            clock.clone(),
+            cancellation_token.clone(),
+            error_state.clone(),
+            fp_registry.clone(),
+        );
+        let join = tokio::spawn(async move { dispatcher.run().await });
+
+        fail_parallel::cfg(fp_registry.clone(), "dispatcher-run-loop", "pause").unwrap();
+        tx.send(TestMessage::Channel(99)).unwrap();
+        clock.advance(Duration::from_millis(5)).await;
+        tx.send(TestMessage::Channel(100)).unwrap();
+        fail_parallel::cfg(fp_registry.clone(), "dispatcher-run-loop", "off").unwrap();
+        wait_for_message_count(log.clone(), 3).await;
+
+        // Shutdown cleanly
+        cancellation_token.cancel();
+        let result = timeout(Duration::from_secs(30), join)
+            .await
+            .expect("dispatcher did not stop in time")
+            .expect("join failed");
+
+        // Verify final state
+        assert!(matches!(result, Err(SlateDBError::BackgroundTaskShutdown)));
+        assert!(matches!(
+            error_state.reader().read(),
+            Some(SlateDBError::BackgroundTaskShutdown)
+        ));
+        assert!(matches!(
+            cleanup_called.reader().read(),
+            Some(Err(SlateDBError::BackgroundTaskShutdown))
+        ));
+        let messages = log.lock().unwrap().clone();
+        assert_eq!(
+            messages,
+            vec![
+                (Phase::Pre, TestMessage::Channel(99)),
+                (Phase::Pre, TestMessage::Channel(100)),
+                (Phase::Pre, TestMessage::Tick(1)),
+            ]
+        );
+    }
+
+    #[cfg(feature = "test-util")]
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn test_dispatcher_supports_multiple_tickers() {
+        let log = Arc::new(Mutex::new(Vec::<(Phase, TestMessage)>::new()));
+        let cleanup_called = WatchableOnceCell::new();
+        let (_tx, rx) = mpsc::unbounded_channel::<TestMessage>();
+        let handler = TestHandler::new(log.clone(), cleanup_called.clone())
+            .add_ticker(Duration::from_millis(5), 1)
+            .add_ticker(Duration::from_millis(7), 2);
+        let clock = Arc::new(MockSystemClock::new());
+        let cancellation_token = CancellationToken::new();
+        let error_state = WatchableOnceCell::new();
+        let mut dispatcher = MessageDispatcher::new(
+            Box::new(handler),
+            rx,
+            clock.clone(),
+            cancellation_token.clone(),
+            error_state.clone(),
+        );
+        let join = tokio::spawn(async move { dispatcher.run().await });
+
+        assert_eq!(log.lock().unwrap().clone(), vec![]);
+        clock.advance(Duration::from_millis(5)).await;
```

Comment:
should we test with two tickers that overlaps (two go off on the same tick?) e.g. [3,5] would overlap at ts=15

Reply:
Will do.

Reply:
Done! And I added one for identical ticks as well.